File: cpp/Mod_eval.cpp
#pragma once

Tuple2<Tree, Map<Node, String>> eval(Tuple2<Tree, Map<Node, String>> ast){
	Map<Node, String> map = get22(ast);
	Tree tree = get21(ast);
	
	String trueS = "TRUE";
	String falseS = "FALSE";
	String notS = "NOT";
	String andS = "AND";
	String orS = "OR";
	String implyS = "IMPLY";
	
	bool done = 0;
	
	while(!done){
		done = 1;
		
		List<Node> nots = inverseOf(map, notS);

		for(int i = 0; i<length(nots); i++){

			Node notNode = get(nots, i);
			Node child = get(children(tree, notNode), 0);
			if(!isLeaf(tree, child))
				continue;
			if(imageOf(map, child) == trueS){
				done = 0;
				tree = removeLeaf(tree, child);
				map = setImage(map, notNode, falseS);
				map = undefine(map,child);
			}else if (imageOf(map, child) == falseS){
				done = 0;
				tree = removeLeaf(tree, child);
				map = setImage(map, notNode, trueS);
				map = undefine(map,child);
			}
		}

		List<Node> ands = inverseOf(map, andS);

		for(int i = 0; i<length(ands); i++){
			Node andNode = get(ands, i);
			Node child1 = get(children(tree, andNode), 0);
			Node child2 = get(children(tree, andNode), 1);
			if(!isLeaf(tree, child1) || !isLeaf(tree, child2))
				continue;
			if(imageOf(map, child1) == trueS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, andNode, imageOf(map, child2));
				map = undefine(map,child1);
				map = undefine(map,child2);
			}else if (imageOf(map, child1) == falseS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, andNode, falseS);
				map = undefine(map,child1);
				map = undefine(map,child2);
			}else if (imageOf(map, child2) == trueS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, andNode, imageOf(map, child1));
				map = undefine(map,child1);
				map = undefine(map,child2);
			}else if (imageOf(map, child2) == falseS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, andNode, falseS);
				map = undefine(map,child1);
				map = undefine(map,child2);
			}

		}


		List<Node> ors = inverseOf(map, orS);

		for(int i = 0; i<length(ors); i++){
			Node orNode = get(ors, i);
			Node child1 = get(children(tree, orNode), 0);
			Node child2 = get(children(tree, orNode), 1);
			if(!isLeaf(tree, child1) || !isLeaf(tree, child2))
				continue;
			if(imageOf(map, child1) == trueS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, orNode, trueS);
				map = undefine(map,child1);
				map = undefine(map,child2);
			}else if (imageOf(map, child1) == falseS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, orNode, imageOf(map, child2));
				map = undefine(map,child1);
				map = undefine(map,child2);
			}else if (imageOf(map, child2) == trueS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, orNode, trueS);
				map = undefine(map,child1);
				map = undefine(map,child2);
			}else if (imageOf(map, child2) == falseS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, orNode, imageOf(map, child1));
				map = undefine(map,child1);
				map = undefine(map,child2);
			}
		}


		List<Node> implys = inverseOf(map, implyS);

		for(int i = 0; i<length(implys); i++){
			Node implyNode = get(implys, i);
			Node child1 = get(children(tree, implyNode), 0);
			Node child2 = get(children(tree, implyNode), 1);
			if(!isLeaf(tree, child1) || !isLeaf(tree, child2))
				continue;
			if(imageOf(map, child1) == trueS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, implyNode, imageOf(map, child2));
				map = undefine(map,child1);
				map = undefine(map,child2);
			}else if (imageOf(map, child1) == falseS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, implyNode, trueS);
				map = undefine(map,child1);
				map = undefine(map,child2);
			}else if (imageOf(map, child2) == trueS){
				done = 0;
				tree = removeLeaf(tree, child1);
				tree = removeLeaf(tree, child2);
				map = setImage(map, implyNode, trueS);
				map = undefine(map,child1);
				map = undefine(map,child2);
			}else if (imageOf(map, child2) == falseS){
				done = 0;
				tree = removeLeaf(tree, child2);
				map = setImage(map, implyNode, notS);
				map = undefine(map,child2);
			}
		}
		
	}

	return tuple2(tree, map);
}

File: hs/Mod_eval.hs
module Mod_eval ( eval ) where
import Prelude hiding (and, or)
import Debug.Trace
import Splib
import Types

eval :: (Tree, (Map Node String)) -> (Tree, (Map Node String))

eval ast = evalHelper (get22 ast) (get21 ast) 0

evalHelper mapParam tree done
 | done == 1 = (tree, mapParam)
 | otherwise = evalHelper (get33 implyfertig) (get32 implyfertig) (get31 implyfertig)
 where
  nots = inverseOf mapParam "NOT"
  ands = inverseOf mapParam "AND"
  ors = inverseOf mapParam "OR"
  implys = inverseOf mapParam "IMPLY"
  notfertig = checkNots mapParam tree nots 0 1
  andfertig = checkAnds (get33 notfertig) (get32 notfertig) ands 0 (get31 notfertig)
  orfertig = checkOrs (get33 andfertig) (get32 andfertig) ors 0 (get31 andfertig)
  implyfertig = checkImplys (get33 orfertig) (get32 orfertig) implys 0 (get31 orfertig)
 
checkNots mapParam tree nots i done
 | i == length nots = (done, tree, mapParam)
 | sp_not (isLeaf tree child) = checkNots mapParam tree nots (i+1) done
 | imageOf mapParam child == "TRUE" = checkNots (undefine (setImage mapParam notNode "FALSE") child) (removeLeaf tree child) nots (i+1) 0
 | imageOf mapParam child == "FALSE" = checkNots (undefine (setImage mapParam notNode "TRUE") child) (removeLeaf tree child) nots (i+1) 0
 | otherwise = checkNots mapParam tree nots (i+1) done
 where 
  notNode = get nots i
  child = get (children tree notNode) 0
  
checkAnds mapParam tree nodes i done
 | i == length nodes = (done, tree, mapParam)
 | sp_not (isLeaf tree child1) || sp_not (isLeaf tree child2) = checkAnds mapParam tree nodes (i+1) done
 | imageOf mapParam child1 == "TRUE" = checkAnds (undefine (undefine (setImage mapParam node (imageOf mapParam child2)) child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | imageOf mapParam child1 == "FALSE" = checkAnds (undefine (undefine (setImage mapParam node "FALSE") child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | imageOf mapParam child2 == "TRUE" = checkAnds (undefine (undefine (setImage mapParam node (imageOf mapParam child1)) child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | imageOf mapParam child2 == "FALSE" = checkAnds (undefine (undefine (setImage mapParam node "FALSE") child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | otherwise = checkAnds mapParam tree nodes (i+1) done
 where 
  node = get nodes i
  child1 = get (children tree node) 0
  child2 = get (children tree node) 1
  
  
  
checkOrs mapParam tree nodes i done
 | i == length nodes = (done, tree, mapParam)
 | sp_not (isLeaf tree child1) || sp_not (isLeaf tree child2) = checkOrs mapParam tree nodes (i+1) done
 | imageOf mapParam child1 == "TRUE" = checkOrs (undefine (undefine (setImage mapParam node "TRUE") child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | imageOf mapParam child1 == "FALSE" = checkOrs (undefine (undefine (setImage mapParam node (imageOf mapParam child2)) child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | imageOf mapParam child2 == "TRUE" = checkOrs (undefine (undefine (setImage mapParam node "TRUE") child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | imageOf mapParam child2 == "FALSE" = checkOrs (undefine (undefine (setImage mapParam node (imageOf mapParam child1)) child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | otherwise = checkOrs mapParam tree nodes (i+1) done
 where 
  node = get nodes i
  child1 = get (children tree node) 0
  child2 = get (children tree node) 1
  
  
checkImplys mapParam tree nodes i done
 | i == length nodes = (done, tree, mapParam)
 | sp_not (isLeaf tree child1) || sp_not (isLeaf tree child2) = checkImplys mapParam tree nodes (i+1) done
 | imageOf mapParam child1 == "TRUE" = checkImplys (undefine (undefine (setImage mapParam node (imageOf mapParam child2)) child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | imageOf mapParam child1 == "FALSE" = checkImplys (undefine (undefine (setImage mapParam node "TRUE") child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | imageOf mapParam child2 == "TRUE" = checkImplys (undefine (undefine (setImage mapParam node "TRUE") child1) child2) (removeLeaf (removeLeaf tree child1) child2) nodes (i+1) 0
 | imageOf mapParam child2 == "FALSE" = checkImplys (undefine (setImage mapParam node "NOT") child2) (removeLeaf tree child2) nodes (i+1) 0
 | otherwise = checkImplys mapParam tree nodes (i+1) done
 where 
  node = get nodes i
  child1 = get (children tree node) 0
  child2 = get (children tree node) 1